blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractMetricsRW.java

```diff
@@ -0,0 +1,324 @@
+        // in this loop, we will fire all the executeAsync() of
+        // various select statements, the collect all of the
+        // ResultSetFutures
+        Map<Locator, List<ResultSetFuture>> locatorToFuturesMap = new HashMap<Locator, List<ResultSetFuture>>();
+        Map<Locator, DAbstractMetricIO> locatorIOMap = new HashMap<Locator, DAbstractMetricIO>();
+
+        for (Locator locator : locators) {
+            try {
+
+
+                String type = metadataCache.get( locator, MetricMetadata.TYPE.name().toLowerCase() );
+                String rType = metadataCache.get( locator, MetricMetadata.ROLLUP_TYPE.name().toLowerCase() );
+
+                DAbstractMetricIO io = getIO( locatorIOMap, locator, type, rType, granularity );
+
+                // do the query
+                List<ResultSetFuture> selectFutures = io.selectForLocatorAndRange( columnFamily, locator, range );
+
+                // add all ResultSetFutures for a particular locator together
+                List<ResultSetFuture> existing = locatorToFuturesMap.get(locator);
+                if (existing == null) {
+                    existing = new ArrayList<ResultSetFuture>();
+                    locatorToFuturesMap.put(locator, existing);
+                }
+                existing.addAll(selectFutures);
+
+            } catch (CacheException ex) {
+                LOG.error(String.format("Error looking up locator %s in cache", locator), ex);
+            }
+        }
+        return resultSetsToMetricData(locatorToFuturesMap, locatorIOMap, granularity);
+    }
+
+    /**
+     * Given a locator, a data type, a rollup type and granularity, return the correct
+     * {@link com.rackspacecloud.blueflood.io.datastax.DAbstractMetricIO} object
+     *
+     * @param locatorIOMap the Locator -> DabstractMetricIO map to add the new returned object to as a side-effect
+     * @param locator
+     * @param dataType
+     * @param rollupType
+     * @param gran
+     * @return
+     */
+    private DAbstractMetricIO getIO( Map<Locator, DAbstractMetricIO> locatorIOMap,
+                                     Locator locator,
+                                     String dataType,
+                                     String rollupType,
+                                     Granularity gran ) {
+
+        DAbstractMetricIO io;
+
+        if( rollupType == null && gran == Granularity.FULL ) {
+
+            io = simpleIO;
+        }
+        else {
+
+            // find out the rollupType for this locator
+            RollupType rType = RollupType.fromString( rollupType );
+
+            // get the right PreaggregatedIO class that can process
+            // this rollupType
+            io = rollupTypeToIO.get(rType);
+        }
+
+
+        if (io == null) {
+            throw new IllegalArgumentException(String.format("dont know how to handle locator=%s with rollupType=%s and type=%s", locator, rollupType, dataType));
```

Comment:
I am not sure I understand your reply here. My comment was for the exception that is thrown. Here, when we get null io, we throw IllegalArgumentException. In line 78, we throw InvalidDataException. Should we be making them throw the same exception?